### PR TITLE
Atomics

### PR DESCRIPTION
**Pull request by paulidale**
- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
If/when we get infallible locking, the ref count APIs should be updated to not return a success code.

File: crypto/asn1/tasn_utl.c [hunk 1]
```diff
@@ -68,8 +68,8 @@ int ossl_asn1_set_choice_selector(ASN1_VALUE **pval, int value,
 int ossl_asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
 {
     const ASN1_AUX *aux;
-    CRYPTO_REF_COUNT *lck;
     CRYPTO_RWLOCK **lock;
```
Comment (kroeckx) [inline crypto/asn1/tasn_utl.c:71]: lock, including creating and freeing it can be removed.

Comment (kroeckx) [reply crypto/asn1/tasn_utl.c:71]: including the ref_lock in the ASN1_AUX structure.

Comment (paulidale) [reply crypto/asn1/tasn_utl.c:71]: It cannot.  It's used for other purposes.
This PR avoids using it for the reference count only.

There are other similar instances.

File: include/internal/refcount.h [hunk 2]
```diff
@@ -25,12 +25,13 @@
 
 #   define HAVE_ATOMICS 1
 
-typedef _Atomic int CRYPTO_REF_COUNT;
+typedef struct {
+    _Atomic int val;
+} CRYPTO_REF_COUNT;
```
Comment (kroeckx) [inline include/internal/refcount.h:30]: Is there a reason this has to be changed to a struct? I currently don't see why the implementations that have atomic support would require a struct. I assume the compiler will actually generate the same code for both.

Comment (paulidale) [reply include/internal/refcount.h:30]: I did this for a couple of reasons:
- There is one flavour that requires a lock and hence must use a structure.  Making all of the different versions use a structure maintains consistency.   I take advantage of this right down the bottom of the files, however changing those to work either way would be straightforward.
- Freedom to change things, it's an opaque struct so we can make changes in the future if required.

I agree that the compiler is going to generate the same code regardless.

**Review by t8m — APPROVED** [review]
Nice work!

**Comment by beldmit** [comment]
Nice work!

JFYI. With and without this patch the test suite runs the same time on x86_64. Do we expect any performance changes? I didn't run @mattcaswell 's tests, BTW

**Comment by t8m** [comment]
> Nice work!
> 
> JFYI. With and without this patch the test suite runs the same time on x86_64. Do we expect any performance changes? I didn't run @mattcaswell 's tests, BTW

IMO no perf changes are expected from this. This is just refactoring (except for those removed lock allocations, which I assume have just negligible impact). On compilers providing well-known atomic calls, they are used already.

But it is technical debt removal.

File: crypto/asn1/tasn_utl.c [hunk 3]
```diff
@@ -78,30 +78,34 @@ int ossl_asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
...
         return 0;
-    lck = offset2ptr(*pval, aux->ref_offset);
     lock = offset2ptr(*pval, aux->ref_lock);
+    refcnt = offset2ptr(*pval, aux->ref_offset);
 
     switch (op) {
     case 0:
-        *lck = ret = 1;
+        if (!CRYPTO_NEW_REF(refcnt, 1))
+            return 0;
```
Comment (kroeckx) [inline crypto/asn1/tasn_utl.c:87]: This should return -1

File: crypto/bio/bio_meth.c [hunk 4]
```diff
@@ -10,26 +10,24 @@
...
-    static CRYPTO_REF_COUNT bio_count = BIO_TYPE_START;
     int newval;
 
     if (!RUN_ONCE(&bio_type_init, do_bio_type_init)) {
         /* Perhaps the error should be raised in do_bio_type_init()? */
         ERR_raise(ERR_LIB_BIO, ERR_R_CRYPTO_LIB);
         return -1;
     }
-    if (!CRYPTO_UP_REF(&bio_count, &newval, bio_type_lock))
+    if (!CRYPTO_UP_REF(&bio_type_count, &newval))
```
Comment (kroeckx) [inline crypto/bio/bio_meth.c:30]: This looks like an abuse of the reference counters that is most likely doing the wrong thing. I don't think this actually guarantees unique numbers, just that it will get increased. Please use CRYPTO_atomic_add().

Comment (kroeckx) [reply crypto/bio/bio_meth.c:30]: There is also no CRYPTO_FREE_REF call.

File: include/internal/refcount.h [hunk 5]
```diff
@@ -193,14 +189,91 @@ static __inline int CRYPTO_GET_REF(volatile int *val, int *ret,
...
+}
+
+static ossl_unused ossl_inline int CRYPTO_GET_REF(CRYPTO_REF_COUNT *refcnt,
+                                                   int *ret)
+{
+    return CRYPTO_atomic_load_int(&refcnt->val, ret, refcnt->lock);
+}
+
+#   define CRYPTO_NEW_FREE_DEFINED  1
+static ossl_unused ossl_inline int CRYPTO_NEW_REF(CRYPTO_REF_COUNT *refcnt, int n)
```
Comment (kroeckx) [inline include/internal/refcount.h:220]: I think the n parameter should get dropped, for a reference count it should always be 1. All cases in the code where it's not 1 seems to be using a wrong API, and should use something like CRYPTO_atomic_add() instead.

File: include/internal/refcount.h [hunk 6]
```diff
@@ -44,106 +45,104 @@ static inline int CRYPTO_UP_REF(_Atomic int *val, int *ret,
...
-    *ret = atomic_fetch_sub_explicit(val, 1, memory_order_relaxed) - 1;
+    *ret = atomic_fetch_sub_explicit(&refcnt->val, 1, memory_order_relaxed) - 1;
     if (*ret == 0)
         atomic_thread_fence(memory_order_acquire);
     return 1;
 }
 
-static inline int CRYPTO_GET_REF(_Atomic int *val, int *ret,
-                                 ossl_unused void *lock)
+static inline int CRYPTO_GET_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
```
Comment (kroeckx) [inline include/internal/refcount.h:56]: This function does not belong in this file. Functions using this is probably yet an other abuse of the API.

**Comment by kroeckx** [comment]
We should probably implement something that does atomics that can be inlined, like refcount.h, but that with acquire + release semantics instead of relaxed and only doing an acquire on the last free. CRYPTO_atomic_add() can than just be a wrapper around that.

File: crypto/bio/bio_meth.c (hunk 4 above)
Comment (kroeckx) [reply crypto/bio/bio_meth.c:30]: It probably does guarantee unique numbers.

File: include/internal/refcount.h (hunk 6 above)
Comment (paulidale) [reply include/internal/refcount.h:56]: Why?  Getting a reference count seems like a natural part of reference counting.

I agree that the API is a bit rubbish.

**Comment by paulidale** [comment]
> We should probably implement something that does atomics that can be inlined, like refcount.h, but that with acquire + release semantics instead of relaxed and only doing an acquire on the last free. CRYPTO_atomic_add() can than just be a wrapper around that.

We're stymied by the fact that our atomics are part of the public API.
The TSAN code is private and does inline.

**Comment by paulidale** [comment]
> Do we expect any performance changes?

No, I'm not expecting any performance change from this.
We will avoid allocating locks unnecessarily but I doubt that's significant.

File: include/internal/refcount.h (hunk 5 above)
Comment (paulidale) [reply include/internal/refcount.h:220]: It is set to zero in `provider_core.c` for the activation count which seems like a legitimate use of a reference counter.  It could be changed to assign 1 and then decrement I suppose.

File: crypto/bio/bio_meth.c (hunk 4 above)
Comment (paulidale) [reply crypto/bio/bio_meth.c:30]: Yes, it does guarantee unique numbers.  There was no free call in the original code.  It allocated the lock and never freed it.  Since this is a pure refactoring, I'd prefer to not address that here.  I've raised an issue: #21277

File: crypto/asn1/tasn_utl.c (hunk 3 above)
Comment (paulidale) [reply crypto/asn1/tasn_utl.c:87]: Updated.  I wasn't sure which to use.

File: crypto/bio/bio_meth.c (hunk 4 above)
Comment (kroeckx) [reply crypto/bio/bio_meth.c:30]: The reason it guarantees a unique number is because the base atomic atomic_fetch_add_explicit() happens to return the old value before incrementing. If it had returned the new value, it would not have been unique. Since this code only calls the CRYPTO_UP_REF, and never the CRYPTO_DOWN_REF it happens to work.

File: include/internal/refcount.h (hunk 5 above)
Comment (kroeckx) [reply include/internal/refcount.h:220]: As mentioned in some other PR, the activation count is not a reference count. It might look like one, but it's not. I have no idea if it's broken or not. If you want to be sure it works properly with relaxed semantics, this will need some careful audit and comments explaining why relaxed is ok and why it's so critical that it's relaxed. I prefer this to be handled with acquire+release semantics. It should be easy to write correct code with acquire+release semantics if you know what you're doing.

Note that x86(_64) does not actually do relaxed.

File: include/internal/refcount.h (hunk 6 above)
Comment (kroeckx) [reply include/internal/refcount.h:56]: Let me rephrase that: This function should not exist.

CRYPTO_UP_REF() should not have returned a number, CRYPTO_DOWN_REF should only indicate that the reference count has become 0 and that you should free it.

CRYPTO_GET_REF(), specially with relaxed semantics, will get you the wrong answer.

Comment (paulidale) [reply include/internal/refcount.h:56]: It's only used in one place `ossl_quic_attach_stream` checking for user error.  This function was supposed to be removed or made internal only if I remember correctly.  @hlandau?  In either case it doesn't need to exist.

Comment (paulidale) [reply include/internal/refcount.h:56]: Raised as  #21278

File: include/internal/refcount.h (hunk 5 above)
Comment (paulidale) [reply include/internal/refcount.h:220]: Raised as #21279

**Comment by paulidale** [comment]
comments?  reviews?

**Review by t8m — APPROVED** [review]

File: crypto/asn1/tasn_utl.c [hunk 7]
```diff
@@ -78,30 +78,34 @@ int ossl_asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
...
         *lock = CRYPTO_THREAD_lock_new();
```
Comment (mattcaswell) [inline crypto/asn1/tasn_utl.c:88]: Is this actually necessary any more? This lock is not used by us anywhere AFAICT. Strictly speaking the "ref_lock" field in the ASN1_AUX structure is public...but are we required to actually initialise it to anything, or can we just set it to NULL?

File: crypto/asn1/tasn_utl.c [hunk 8]
```diff
@@ -78,30 +78,34 @@ int ossl_asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
...
             return -1;
         break;
     case -1:
-        if (!CRYPTO_DOWN_REF(lck, &ret, *lock))
+        if (!CRYPTO_DOWN_REF(refcnt, &ret))
             return -1;  /* failed */
         REF_PRINT_EX(it->sname, ret, (void *)it);
         REF_ASSERT_ISNT(ret < 0);
         if (ret == 0) {
             CRYPTO_THREAD_lock_free(*lock);
```
Comment (mattcaswell) [inline crypto/asn1/tasn_utl.c:106]: As above. Do we actually need to do this?

File: crypto/provider_core.c [hunk 9]
```diff
@@ -1066,7 +1056,7 @@ static int provider_deactivate(OSSL_PROVIDER *prov, int upcalls,
         return -1;
     }
 
-    CRYPTO_atomic_add(&prov->activatecnt, -1, &count, prov->refcnt_lock);
+    CRYPTO_DOWN_REF(&prov->activatecnt, &count);
```
Comment (mattcaswell) [inline crypto/provider_core.c:1059]: In a previous PR we initialy had activatecnt as a refcount, but then deliberately changed it to an int and used `CRYPTO_atomic_add` for reasons described by @kroeckx. Here you are converting it to a ref count again. This doesn't look right. We might need to do the same work for `CRYPTO_atomic_add()` et al, that you are doing here for ref counts. But probably not in this PR. I'd leave refcnt_lock alone (maybe rename it), so that it can be used for the activatecnt and deal with that one in a subsequent PR which converts atomic ints into a structure.

File: ssl/ssl_local.h
Comment (mattcaswell) [inline ssl/ssl_local.h:1]: We've left the lock behind in the ssl structure because its still used in a handful of places. The remaining locations it is used look *highly* suspect to me.

File: crypto/evp/signature.c [hunk 10]
```diff
@@ -25,15 +25,21 @@ static EVP_SIGNATURE *evp_signature_new(OSSL_PROVIDER *prov)
     if (signature == NULL)
         return NULL;
 
+    if (!CRYPTO_NEW_REF(&signature->refcnt, 1)) {
+        OPENSSL_free(signature);
+        return NULL;
+    }
+
     signature->lock = CRYPTO_THREAD_lock_new();
```
Comment (mattcaswell) [inline crypto/evp/signature.c:33]: I don't see anywhere that we still need this lock? Am I missing it?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/asn1/tasn_utl.c (hunk 7 above)
Comment (mattcaswell) [reply crypto/asn1/tasn_utl.c:88]: Yes...in theory...

File: crypto/evp/signature.c (hunk 10 above)
Comment (paulidale) [reply crypto/evp/signature.c:33]: You're right.  I must have missed this one when I went through each removing the lock and checking for failures.

File: crypto/provider_core.c (hunk 9 above)
Comment (paulidale) [reply crypto/provider_core.c:1059]: Fixed.

File: ssl/ssl_local.h
Comment (paulidale) [reply ssl/ssl_local.h:1]: Raised as #21320.

File: crypto/asn1/tasn_utl.c (hunk 7 above)
Comment (paulidale) [reply crypto/asn1/tasn_utl.c:88]: We've maintained things in the name of backward compatibility that were less justified IMO.

Leaving it as NULL would be okay if the locking functions returned when passed a NULL argument.  They don't, they crash.

**Review by mattcaswell — APPROVED** [review]
LGTM.

@t8m - please reconfirm

**Comment by kroeckx** [comment]
I'm trying to review it, will probably have time tomorrow.

**Comment by paulidale** [comment]
@t8m was away yesterday, back today.

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
Please update any docs that refer to these macros (see evp_generic_fetch.pod), Separate PR is ok.

**Comment by paulidale** [comment]
Merged, thanks of the reviews and feedback.

**Comment by kroeckx** [comment]
Note that the 24 hour window didn't pass yet on merge.

File: crypto/bio/bio_lib.c [hunk 11]
```diff
@@ -88,22 +88,17 @@ BIO *BIO_new_ex(OSSL_LIB_CTX *libctx, const BIO_METHOD *method)
...
-        CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
+    if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data))
         goto err;
-    }
 
     if (method->create != NULL && !method->create(bio)) {
         ERR_raise(ERR_LIB_BIO, ERR_R_INIT_FAIL);
         CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-        CRYPTO_THREAD_lock_free(bio->lock);
+        CRYPTO_FREE_REF(&bio->references);
```
Comment (kroeckx) [inline crypto/bio/bio_lib.c:101]: This seems to introduce a double free

File: include/internal/refcount.h [hunk 12]
```diff
@@ -193,14 +189,91 @@ static __inline int CRYPTO_GET_REF(volatile int *val, int *ret,
...
+    if (refcnt->lock == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_CRYPTO_LIB);
+        return 0;
+    }
+    return 1;
+}
+
+static ossl_unused ossl_inline void CRYPTO_FREE_REF(CRYPTO_REF_COUNT *refcnt)                                  \
+{
+    CRYPTO_THREAD_lock_free(refcnt->lock);
```
Comment (kroeckx) [inline include/internal/refcount.h:233]: This function should support refcnt == NULL. The code changes actually make it possible to call this on error with it still being NULL.

File: crypto/dh/dh_lib.c [hunk 13]
```diff
@@ -78,14 +78,16 @@ static DH *dh_new_intern(ENGINE *engine, OSSL_LIB_CTX *libctx)
...
-    ret->references = 1;
     ret->lock = CRYPTO_THREAD_lock_new();
     if (ret->lock == NULL) {
         ERR_raise(ERR_LIB_DH, ERR_R_CRYPTO_LIB);
         OPENSSL_free(ret);
         return NULL;
     }
 
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
```
Comment (kroeckx) [inline crypto/dh/dh_lib.c:89]: You do not want to use goto err; here, it will call DH_free() which wants to do CRYPTO_DOWN_REF().

File: crypto/dsa/dsa_lib.c [hunk 14]
```diff
@@ -137,14 +137,16 @@ static DSA *dsa_new_intern(ENGINE *engine, OSSL_LIB_CTX *libctx)
...
-    ret->references = 1;
     ret->lock = CRYPTO_THREAD_lock_new();
     if (ret->lock == NULL) {
         ERR_raise(ERR_LIB_DSA, ERR_R_CRYPTO_LIB);
         OPENSSL_free(ret);
         return NULL;
     }
 
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
```
Comment (kroeckx) [inline crypto/dsa/dsa_lib.c:148]: Don't use goto err;

File: crypto/ec/ec_kmeth.c [hunk 15]
```diff
@@ -137,6 +133,7 @@ EC_KEY *ossl_ec_key_new_method_int(OSSL_LIB_CTX *libctx, const char *propq,
     return ret;
 
  err:
+    CRYPTO_FREE_REF(&ret->references);
```
Comment (kroeckx) [inline crypto/ec/ec_kmeth.c:136]: This introduces a double free

File: crypto/ec/ec_kmeth.c [hunk 16]
```diff
@@ -93,12 +93,8 @@ EC_KEY *ossl_ec_key_new_method_int(OSSL_LIB_CTX *libctx, const char *propq,
             goto err;
     }
 
-    ret->references = 1;
-    ret->lock = CRYPTO_THREAD_lock_new();
-    if (ret->lock == NULL) {
-        ERR_raise(ERR_LIB_EC, ERR_R_CRYPTO_LIB);
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
         goto err;
```
Comment (kroeckx) [inline crypto/ec/ec_kmeth.c:97]: You do not want to goto err; here

File: crypto/evp/evp_rand.c [hunk 17]
```diff
@@ -66,25 +65,23 @@ static void evp_rand_free(void *vrand)
...
 }
 
 static void *evp_rand_new(void)
 {
     EVP_RAND *rand = OPENSSL_zalloc(sizeof(*rand));
 
-    if (rand == NULL
-            || (rand->refcnt_lock = CRYPTO_THREAD_lock_new()) == NULL) {
+    if (rand == NULL || !CRYPTO_NEW_REF(&rand->refcnt, 1)) {
         OPENSSL_free(rand);
```
Comment (kroeckx) [inline crypto/evp/evp_rand.c:82]: This needlessly calls OPENSSL_free(NULL) in case rand == NULL. The other changes use:
`if (rand != NULL && !CRYPTO_NEW_REF(...`

File: crypto/evp/evp_rand.c [hunk 18]
```diff
@@ -341,15 +338,14 @@ EVP_RAND_CTX *EVP_RAND_CTX_new(EVP_RAND *rand, EVP_RAND_CTX *parent)
     ctx = OPENSSL_zalloc(sizeof(*ctx));
     if (ctx == NULL)
         return NULL;
-    if ((ctx->refcnt_lock = CRYPTO_THREAD_lock_new()) == NULL) {
+    if (!CRYPTO_NEW_REF(&ctx->refcnt, 1)) {
         OPENSSL_free(ctx);
-        ERR_raise(ERR_LIB_EVP, ERR_R_CRYPTO_LIB);
```
Comment (kroeckx) [inline crypto/evp/evp_rand.c:346]: Why remove this?

File: crypto/provider_core.c [hunk 19]
```diff
@@ -442,16 +442,18 @@ static OSSL_PROVIDER *provider_new(const char *name,
 
     if ((prov = OPENSSL_zalloc(sizeof(*prov))) == NULL)
         return NULL;
+    if (!CRYPTO_NEW_REF(&prov->refcnt, 1)) {
+        ossl_provider_free(prov);
```
Comment (kroeckx) [inline crypto/provider_core.c:446]: I think you want OPENSSL_free(prov) here, ossl_provider_free will try to use the refcnt that doesn't exist.

File: crypto/rsa/rsa_lib.c [hunk 20]
```diff
@@ -79,14 +79,16 @@ static RSA *rsa_new_intern(ENGINE *engine, OSSL_LIB_CTX *libctx)
...
-    ret->references = 1;
     ret->lock = CRYPTO_THREAD_lock_new();
     if (ret->lock == NULL) {
         ERR_raise(ERR_LIB_RSA, ERR_R_CRYPTO_LIB);
         OPENSSL_free(ret);
         return NULL;
     }
 
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
```
Comment (kroeckx) [inline crypto/rsa/rsa_lib.c:90]: don't goto err;

File: crypto/x509/x509_lu.c [hunk 21]
```diff
@@ -209,13 +209,16 @@ X509_STORE *X509_STORE_new(void)
...
+
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
     return ret;
 
 err:
     X509_VERIFY_PARAM_free(ret->param);
     sk_X509_OBJECT_free(ret->objs);
     sk_X509_LOOKUP_free(ret->get_cert_methods);
+    CRYPTO_THREAD_lock_free(ret->lock);
```
Comment (kroeckx) [inline crypto/x509/x509_lu.c:221]: This looks like a bugfix that might need backporting.

Comment (kroeckx) [reply crypto/x509/x509_lu.c:221]: Nevermind.

File: ssl/ssl_lib.c [hunk 22]
```diff
@@ -3779,12 +3785,15 @@ SSL_CTX *SSL_CTX_new_ex(OSSL_LIB_CTX *libctx, const char *propq,
...
     }
+
     ret = OPENSSL_zalloc(sizeof(*ret));
     if (ret == NULL)
         goto err;
 
     /* Init the reference counting before any call to SSL_CTX_free */
-    ret->references = 1;
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
```
Comment (kroeckx) [inline ssl/ssl_lib.c:3795]: Don't goto err;

File: ssl/ssl_sess.c [hunk 23]
```diff
@@ -174,13 +171,8 @@ SSL_SESSION *ssl_session_dup(const SSL_SESSION *src, int ticket)
     dest->next = NULL;
     dest->owner = NULL;
 
-    dest->references = 1;
-
-    dest->lock = CRYPTO_THREAD_lock_new();
-    if (dest->lock == NULL) {
-        ERR_raise(ERR_LIB_SSL, ERR_R_CRYPTO_LIB);
+    if (!CRYPTO_NEW_REF(&dest->references, 1))
         goto err;
```
Comment (kroeckx) [inline ssl/ssl_sess.c:175]: Don't goto err;

File: crypto/evp/evp_rand.c (hunk 18 above)
Comment (paulidale) [reply crypto/evp/evp_rand.c:346]: CRYPTO_NEW_REF raises an error on failure, so no need to raise another.

File: ssl/ssl_sess.c (hunk 23 above)
Comment (paulidale) [reply ssl/ssl_sess.c:175]: and properly clean up the initial allocation.

File: include/internal/refcount.h (hunk 12 above)
Comment (paulidale) [reply include/internal/refcount.h:233]: I don't think this is needed.  The argument is a pointer to a field inside a structure and thus cannot ever be null.
Still, the check should be harmless and ought to optimise out.

Comment (kroeckx) [reply include/internal/refcount.h:233]: You're right, it's something that is always already allocated, refcnt should never be NULL.

**Comment by paulidale** [comment]
I've done the fixes in #21341
